File: Binary_Search/Solutions/Separate_Squares_I.py
class Solution:
    def separateSquares(self, squares: List[List[int]]) -> float:
        
        # Step 1: total area of all squares
        total_area = 0
        min_y = float('inf')
        max_y = float('-inf')
        
        for x, y, side in squares:
            total_area += side * side
            min_y = min(min_y, y)
            max_y = max(max_y, y + side)
        
        target = total_area / 2
        
        # Function to compute area below horizontal line y = k
        def area_below(k):
            area = 0
            for x, y, side in squares:
                bottom = y
                top = y + side
                
                if top <= k:
                    area += side * side
                elif bottom < k < top:
                    area += side * (k - bottom)
            return area
        
        # Step 2: Binary search on y-coordinate
        low, high = min_y, max_y
        
        for _ in range(60):  # enough for precision
            mid = (low + high) / 2
            if area_below(mid) < target:
                low = mid
            else:
                high = mid
        
        return low

File: Binary_Search/Solutions/Separate_Squares_I.cpp
class Solution {
public:
    double separateSquares(vector<vector<int>>& squares) {
        
        double totalArea = 0.0;
        double minY = 1e18, maxY = -1e18;

        // Step 1: Calculate total area and vertical bounds
        for (auto &sq : squares) {
            double y = sq[1];
            double side = sq[2];
            totalArea += side * side;
            minY = min(minY, y);
            maxY = max(maxY, y + side);
        }

        double target = totalArea / 2.0;

        // Function to calculate area below y = mid
        auto areaBelow = [&](double mid) {
            double area = 0.0;
            for (auto &sq : squares) {
                double y = sq[1];
                double side = sq[2];
                double bottom = y;
                double top = y + side;

                // Square completely below the line
                if (top <= mid) {
                    area += side * side;
                }
                // Square partially below the line
                else if (bottom < mid) {
                    area += side * (mid - bottom);
                }
                // else: square completely above → contributes 0
            }
            return area;
        };

        // Step 2: Binary search on y-coordinate
        double low = minY, high = maxY;

        // 60 iterations give enough precision for double
        for (int i = 0; i < 60; i++) {
            double mid = (low + high) / 2.0;
            if (areaBelow(mid) < target)
                low = mid;
            else
                high = mid;
        }

        return low;
    }
};